### PR TITLE
std: add BloomFilter data structure

### PR DESCRIPTION
**Pull request by daurnimator**

### CODE REVIEW

File: std/bloom_filter.zig [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+test "std.BloomFilter" {
+    inline for ([_]type{ bool, u1, u2, u3, u4 }) |Cell| {
+        const emptyCell = if (Cell == bool) false else Cell(0);
+        const BF = BloomFilter(128 * 8, 8, Cell, builtin.endian, hashFunc);
+        var bf = BF{};
+        var i: usize = undefined;
+        // confirm that it is initialised to the empty filter
+        i = 0;
+        while (i < BF.items) : (i += 1) {
+            testing.expectEqual(emptyCell, bf.getCell(@intCast(BF.Index, i)));
```
Comment (daurnimator) [inline std/bloom_filter.zig]: This was a place where #2159 would make the code cleaner.

**Comment by daurnimator** [comment]
FYI this PR is blocked on #2850

File: std/bloom_filter.zig [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    st.update(std.mem.asBytes(&Ki));
+    st.update(in);
+    st.final(out);
+}
+
+test "std.BloomFilter" {
+    inline for ([_]type{ bool, u1, u2, u3, u4 }) |Cell| {
+        const emptyCell = if (Cell == bool) false else Cell(0);
+        const BF = BloomFilter(128 * 8, 8, Cell, builtin.endian, hashFunc);
+        var bf = BF{};
```
Comment (shawnl) [inline std/bloom_filter.zig]: What does this line do?

Comment (daurnimator) [reply std/bloom_filter.zig]: Creates an instance of the struct.

**Comment by andrewrk** [comment]
#2850 was closed as "working as designed".

**Comment by daurnimator** [comment]
I managed to figure out a solution using a `packed struct`. This PR should now be mergable.

**Pull request merged**